block/manager.go

```diff
@@ -339,63 +348,90 @@ func (m *Manager) publishBlock(ctx context.Context) error {
 		lastHeaderHash = lastBlock.Header.Hash()
 	}
 
-	m.logger.Info("Creating and publishing block", "height", newHeight)
+	var block *types.Block
 
-	block := m.executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, m.lastState)
-	m.logger.Debug("block info", "num_tx", len(block.Data.Txs))
-	newState, responses, _, err := m.executor.ApplyBlock(ctx, m.lastState, block)
-	if err != nil {
-		return err
+	// Check if there's an already stored block at a newer height
+	// If there is use that instead of creating a new block
+	pendingBlock, err := m.store.LoadBlock(newHeight)
+	if err == nil {
+		m.logger.Info("Using pending block", "height", newHeight)
+		block = pendingBlock
+	} else {
+		m.logger.Info("Creating and publishing block", "height", newHeight)
+		block = m.executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, m.lastState)
+		m.logger.Debug("block info", "num_tx", len(block.Data.Txs))
+
+		headerBytes, err := block.Header.MarshalBinary()
+		if err != nil {
+			return err
+		}
+		sign, err := m.proposerKey.Sign(headerBytes)
+		if err != nil {
+			return err
+		}
+		commit := &types.Commit{
+			Height:     block.Header.Height,
+			HeaderHash: block.Header.Hash(),
+			Signatures: []types.Signature{sign},
+		}
+
+		// SaveBlock commits the DB tx
+		err = m.store.SaveBlock(block, commit)
+		if err != nil {
+			return err
+		}
 	}
 
-	headerBytes, err := block.Header.MarshalBinary()
+	// Apply the block but DONT commit
+	newState, responses, err := m.executor.ApplyBlock(ctx, m.lastState, block)
 	if err != nil {
 		return err
 	}
-	sign, err := m.proposerKey.Sign(headerBytes)
+
+	err = m.submitBlockToDA(ctx, block)
 	if err != nil {
+		m.logger.Error("Failed to submit block to DA Layer")
 		return err
 	}
 
-	commit := &types.Commit{
-		Height:     block.Header.Height,
-		HeaderHash: block.Header.Hash(),
-		Signatures: []types.Signature{sign},
-	}
-	err = m.store.SaveBlock(block, commit)
+	// Only update the stored height after successfully submitting to DA layer
+	m.store.SetHeight(block.Header.Height)
```

Comment:
What happens if `submitBlockToDA` was successful but the node crashes immediately after that? Can it still recover? What about the other steps in between?

Reply:
I'll admit I'm not 100% sure. 

The `BlockResponse`, `State`, and `Validators` are never read from `Store` so those calls don't currently matter. 

If it crashes before `m.store.SetHeight(block.Header.Height)` or `m.executor.Commit(ctx, newState, block, responses)` are called (which is immediately after `submitBlockToDA()`) then there may be an issue. 

The sequencer will have written a block to the DA layer but *will not* have broadcast a header for other nodes to retrieve this block. Since state hasn't been committed to the `proxyApp` (via `m.executor.Commit()`) I _think_  the sequencer will simply try again with the same block as  the block height hasn't been incremented.

I think the primary consequence here is that the sequencer will have to pay Celestia to submit a block multiple times. This could be handled better by having the sequencer check if the block is already present in the DA layer by using `fetchBlock(daHeight)`. 

I'll create an issue to track this but would like to leave it for a later PR.